enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,407 @@
+* Unfortunately, some kernel tuning needs to be defined at boot time and is static meaning the same tuning is
+applied to all application CPUs whether it is required or not.
+* Some of these static tunings are problematic for workload 2, such as nohz_full and rcu callbacks. So we
+need the tuning for workload 3 but it is problematic for workload 2.
+
+This proposal provides an option for creating a separate pre-defined pool of CPUs for workload 2 (i.e. 
+shared CPUs) and workload 3 (i.e guaranteed CPUs), which will allow the necessary kernel tuning to be 
+applied to workload 3 without adversely affecting workload 2.
+
+### User Stories
+
+* As a telco service provider, I want to configure the set of CPUs running the L1/L2 functions of a vDU 
+application for extremely low latency, without impacting the other application components running on the 
+same server (on different CPUs).
+
+### Goals
+
+* Allow the configuration of separate pre-defined pools of shared and guaranteed CPUs, with different
+kernel tuning applied to each pool.
+* Automatically assign containers that use whole CPUs and are in a pod with the guaranteed QoS class to the
+guaranteed CPUs.
+* The selection of shared vs. guaranteed CPUs should be transparent to the user - no changes to the pod
+spec should be required.
+* The configuration should be optional and activated with a feature gate.
+* Although the initial usecase for this enhancement is single-node deployments, it should also be usable
+for multi-node deployments.
+
+### Non-Goals
+
+* This enhancement assumes the feature is activated as part of installing the cluster and cannot be
+activated later. However, the sizing of the pools of shared and guaranteed CPUs can be modified (a reboot
+will be required).
+* The use of the shared and guaranteed CPU pools will be activated at the cluster level - it cannot be
+configured for a subset of the nodes in a cluster.
+
+## Proposal
+
+The [cluster-node-tuning-operator (NTO)](https://github.com/openshift/cluster-node-tuning-operator) will be
+extended to add a shared CPUSet to the [Performance Profile CPU configuration](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#cpu)
+in addition to the existing reserved and isolated CPUSets. For example:
+
+```yaml
+cpu:
+  reserved: 0,1
+  shared: 2-5
+  isolated: 6-15
+```
+When the shared CPUSet is configured, NTO will:
+* Continue to set the systemd.cpu_affinity to the reserved CPUSet.
+* Apply kernel configuration for isolated CPUs (e.g. nohz_full) to only the isolated CPUSet.
+* Update kubelet configuration to specify that the isolated CPUSet is to be used for guaranteed QoS
+containers using whole CPUs.
+* Update OpenShift Kubernetes API Server configuration to enable a new admission hook. Note: although this would
+work in single-node deployments, we still need to work through how enablement will work in multi-node deployments.
```

Comment:
So as for now it should not work in MNO cluster?

Reply:
I think we should figure this out before the enhancement is completed. I wanted to get some general feedback on the approach before sorting out all the details.

Reply:
There should be no reason this can't work on a MNO

Reply:
Clarified that we will also consider multi-node deployments.